controllers/elasticdl/status.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2019 The Alibaba Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package elasticdl
+
+import (
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	elasticdlv1alpha1 "github.com/alibaba/kubedl/api/elasticdljob/v1alpha1"
+	v1 "github.com/alibaba/kubedl/pkg/job_controller/api/v1"
+	commonutil "github.com/alibaba/kubedl/pkg/util"
+)
+
+// updateGeneralJobStatus updates the status of job with given replica specs and job status.
+func (r *ElasticDLJobReconciler) updateGeneralJobStatus(elasticdlJob *elasticdlv1alpha1.ElasticDLJob,
+	replicaSpecs map[v1.ReplicaType]*v1.ReplicaSpec, jobStatus *v1.JobStatus, restart bool) error {
+	log.Info("Updating status", "ElasticDLJob name", elasticdlJob.Name, "restart", restart)
+
+	// Set job status start time since this job has acknowledged by controller.
+	if jobStatus.StartTime == nil {
+		now := metav1.Now()
+		jobStatus.StartTime = &now
+	}
+
+	previousRestarting := commonutil.IsRestarting(*jobStatus)
+	previousFailed := commonutil.IsFailed(*jobStatus)
+
+	for rtype, spec := range replicaSpecs {
+		replicas := *spec.Replicas
+		// If rtype in replica status not found, there must be a mistyped/invalid rtype in job spec,
+		// and it has not been reconciled in previous processes, discard it.
+		status, ok := jobStatus.ReplicaStatuses[rtype]
+		if !ok {
+			log.Info("skipping invalid replica type", "rtype", rtype)
+			continue
+		}
+		expected := replicas - status.Succeeded
+		running := status.Active
+		failed := status.Failed
+
+		log.Info("Update elasticdl job status", "ElasticDLJob", elasticdlJob.Name,
+			"ReplicaType", rtype, "expected", expected, "running", running, "failed", failed)
+
+		if ContainMasterSpec(elasticdlJob) {
```

Comment:
should we move this condition checking out of the for-loop?

Reply:
Good idea. Since an elasticdl job only contains a master replica spec, we could remove the for-loop logic.